arangod/Cluster/DBServerAgencySync.cpp

```diff
@@ -52,125 +60,176 @@ void DBServerAgencySync::work() {
+
+  using namespace arangodb::basics;
+  Result result;
+  DatabaseFeature* dbfeature = nullptr;
+
+  try {
+    dbfeature = ApplicationServer::getFeature<DatabaseFeature>("Database");
+  } catch (...) {}
+
+  if (dbfeature == nullptr) {
+    LOG_TOPIC(ERR, Logger::HEARTBEAT) << "Failed to get feature database";
+    return Result(TRI_ERROR_INTERNAL, "Failed to get feature database");
+  }
+
+  collections.clear();
+  VPackObjectBuilder c(&collections);
+  for (auto const& database : Databases::list()) {
+
+    try {
+      DatabaseGuard guard(database);
+      auto vocbase = &guard.database();
+      
+      collections.add(VPackValue(database));
+      VPackObjectBuilder db(&collections);
+      auto cols = vocbase->collections(false);
+      
+      for (auto const& collection : cols) {
+        collections.add(VPackValue(collection->name()));
+        VPackObjectBuilder col(&collections);
+        collection->toVelocyPack(collections,true,false);
+        collections.add(
+          "theLeader", VPackValue(collection->followers()->getLeader()));
+      }        
+    } catch (std::exception const& e) {
+      return Result(
+        TRI_ERROR_INTERNAL,
+        std::string("Failed to guard database ") +  database + ": " + e.what());
+    }
+    
+  }
+  
+  return Result();
+  
+}
+
 DBServerAgencySyncResult DBServerAgencySync::execute() {
   // default to system database
 
-  LOG_TOPIC(DEBUG, Logger::HEARTBEAT) << "DBServerAgencySync::execute starting";
-  DatabaseFeature* database = 
+  TRI_ASSERT(AgencyCommManager::isEnabled());
+  AgencyComm comm;
+  
+  using namespace std::chrono;
+  using clock = std::chrono::steady_clock;
+  
+  LOG_TOPIC(DEBUG, Logger::MAINTENANCE) << "DBServerAgencySync::execute starting";
+  DatabaseFeature* dbfeature = 
     ApplicationServer::getFeature<DatabaseFeature>("Database");
-  TRI_vocbase_t* const vocbase = database->systemDatabase();
+  MaintenanceFeature* mfeature = 
+    ApplicationServer::getFeature<MaintenanceFeature>("Maintenance");
+  TRI_vocbase_t* const vocbase = dbfeature->systemDatabase();
+  
   DBServerAgencySyncResult result;
 
   if (vocbase == nullptr) {
-    LOG_TOPIC(DEBUG, Logger::HEARTBEAT)
+    LOG_TOPIC(DEBUG, Logger::MAINTENANCE)
       << "DBServerAgencySync::execute no vocbase";
     return result;
   }
   
+  Result tmp;
+  VPackBuilder rb;
   auto clusterInfo = ClusterInfo::instance();
   auto plan = clusterInfo->getPlan();
-  auto current = clusterInfo->getCurrent();
-  DatabaseGuard guard(*vocbase);
-  double startTime = TRI_microtime();
-  V8Context* context = V8DealerFeature::DEALER->enterContext(vocbase, true, V8DealerFeature::ANY_CONTEXT_OR_PRIORITY);
+  auto serverId = arangodb::ServerState::instance()->getId();
 
-  if (context == nullptr) {
-    LOG_TOPIC(WARN, arangodb::Logger::HEARTBEAT) << "DBServerAgencySync::execute: no V8 context";
+  VPackBuilder local;
+  Result glc = getLocalCollections(local);
+  if (!glc.ok()) {
```

Comment:
if getLocalCollections() fails, we return "result" which says everything is "ok".  Is that true?  I also notice that there is another Result object "tmp" that is going to eat bad results in few lines later.  Again, is that the correct action?  If so, how about supporting comments in the code for both.

Reply:
Well spotted, added a FIXMEMAINTENANCE for post-merge changes.